TD-2263 Adding infofile validator

arpadlomenipg opened the pull request:
<!-- Insert YouTrack link if relevant -->

Contributes to[ TD-2263](https://sce.myjetbrains.com/youtrack/issue/TD-2263/VIRTO.TEST-errors-when-Vehicle-folder-contains-temp-files)

## Changes

A supporting function for [ TD-2263](https://sce.myjetbrains.com/youtrack/issue/TD-2263/VIRTO.TEST-errors-when-Vehicle-folder-contains-temp-files), but could be used in other applications too later.

It adds an infofile validator to the infofile-utils. Additionally, I infofiles into the repository for the unit tests. The current options to for infofile validations are : 
* "Vehicle" (looks either for Car, Motorcycle or Truck infofile)
* "TestRun" (looks for TestRun infofile)
* "Road" (looks for Road infofile)
* "Model" (looks for anything else, which is a valid infofile, be that a Trailer, a sensor model, etc.

The basis of the validation is always the "FileIdent" property of the infofile, an appropriate RegEx search is carried out for each of the options listed above.

## Dependencies

## UI/UX

## Testing notes

## Author checklist before assigning a reviewer

- [x] Reviewed my own code-diff.
- [x] Branch has been run in docker.
- [x] PR assigned to me or an appropriate delegate.
- [x] Relevant labels added to the PR.
- [x] Appropriate tests have been added.
- [x] Lint and test workflows pass.

mattcorner commented on src/isValidInfoFile.js:
```diff
@@ -0,0 +1,120 @@
+    switch (type) {
+      // regex match for possible vehicle types in the FileIdent property of the infofile
+      case "Vehicle":
+        if (
+          fileIdent.match(/^CarMaker-Car\s\d{2}$/) ||
+          fileIdent.match(/^CarMaker-Truck\s\d{2}$/) ||
+          fileIdent.match(/^CarMaker-Motorcycle\s\d{2}$/)
+        ) {
+          // delete the infofile handle
+          infofile.delete();
+          return true;
+        } else {
+          // delete the infofile handle
+          infofile.delete();
+          return false;
+        }
```
Generally speaking, if else statements already operate on a boolean, so there is no need to return booleans from within. Instead just return the statement. Same applies to the rest of the switch cases.

```suggestion
        return fileIdent.match(/^CarMaker-Car\s\d{2}$/) ||
          fileIdent.match(/^CarMaker-Truck\s\d{2}$/) ||
          fileIdent.match(/^CarMaker-Motorcycle\s\d{2}$/)
```

Granted here you are also handling infofile.delete() which I'll comment on seperately.

arpadlomenipg replied:
The if-else statements are operating on vectors (.match returns that), so this solution cannot be used here. However, with the elimination of all the infofile.delete-s, I can use the ? : operator now, which will make the code a bit more compact

mattcorner replied:
Makes sense

mattcorner commented on src/isValidInfoFile.js:
```diff
@@ -0,0 +1,120 @@
+
+    // throw the error
+    throw new Error(error);
+  }
```
You basically always want to delete the infofile handle before returning the result or an error. Hence your code is littered with calls to `infofile.delete()`. There is a third step to a `try {} catch {}` called finally which handles this exact case. It runs regardless of whether we return a result or throw an error.

Hence you can do:
```ts
try {
  // ...
} catch (err) {
   // ...
} finally {
   infofile.delete();
}

mattcorner commented on src/index.js:
```diff
@@ -13,4 +13,5 @@ module.exports = {
   ...require("./setDouble"),
   ...require("./setText"),
   ...require("./setValue"),
+  ...require("./isValidInfoFile"),
```
We should update the README.md now that we are exporting an additional method from the package.

arpadlomenipg replied:
updated the README, thanks for the feedback!

mattcorner commented on src/__tests__/isValidInfoFile.test.js:
```diff
@@ -0,0 +1,151 @@
+// get double tests
+describe("isValidInfoFile tests", () => {
+  // test case validates a valid car infofile successfully with the right function call
+  test("validates a valid car infofile successfully with the right function call", () => {
+    expect(
+      isValidInfoFile({
+        file: fileCar,
+        type: "Vehicle",
+      })
+    ).toEqual(true);
+  });
```
There are a lot of similar tests in this file that basically say "For this file and infofile type, check that the result is true".

Check out parameterised tests using [`test.each`](https://vitest.dev/api/#test-each). You basically create a matrix of test data, and then write one test.

mattcorner commented on src/isValidInfoFile.js:
```diff
@@ -0,0 +1,120 @@
+    throw new Error("type is required");
+  }
+
+  if (!isValidType(type)) {
+    throw new Error("type is invalid");
+  }
```
I expected to be able to call this method without a infofile type. a.k.a I just want to check its an infofile but I don't care what type. I think we did discuss this but I guess it got lost in the detail.

arpadlomenipg replied:
I must have missed that then, added this to the code

mattcorner commented on src/__tests__/isValidInfoFile.test.js:
```diff
@@ -0,0 +1,151 @@
+  });
+
+  // test case returns false when the provided file is not a valid infofile
+  test("returns false when the provided file is not a valid infofile", () => {
+    expect(
+      isValidInfoFile({
+        file: tempFile,
+        type: "TestRun",
+      })
+    ).toEqual(false);
+  });
```
Can you describe the path through the code that gets us here? Reading the code it looks like it would error. 

I presume the infofile handle does construct, and get FileIdent actually returns an empty string or something?

arpadlomenipg replied:
yes, that's exactly what happens

mattcorner reviewed (changes requested):
Nice work, left some suggestions.

mattcorner commented on README.md:
```diff
@@ -260,13 +260,45 @@ console.log(status)
+const infofile = require("@ipguk/infofile-utils");
+
+// get the absolute path to the infofile
+const file = C:\infofiles\infofile
```
```suggestion
const file = "C:\infofiles\infofile";
```

arpadlomenipg replied:
I will fix this in all other places of the README too

mattcorner reviewed (changes requested):
Minor typo in the README.

mattcorner commented:
I've published a pre-release for you so that you can start making use if it in VIRTO as a final test: @ipguk/infofile-utils@1.1.0-0

mattcorner reviewed (changes requested):
failing CI

arpadlomenipg commented:
Sorry, turns out file names are case insensitive on windows, but not on linux, hence the unexpected unit test fail here

I added a bunch of new types, removed the type "Model", changed the RegEx search pattern when calling without type

arpadlomenipg commented:
I am done with the testing in Virto, I have not encountered any problems. I looked through the code, this implementation has all the necessary types to use it in Virto-apps, where currently an infofile validation is being used with a local implementation, so I have no plans to expand it with more types now.

mattcorner reviewed (approved):
lgtm. merge the PR and drop me a message to do the release
